src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,880 @@
+                    source_info: callsite.location,
+                    kind: TerminatorKind::Goto { target: BasicBlock::new(bb_len) }
+                };
+
+                caller_mir[callsite.bb].terminator = Some(terminator);
+
+                if let Some(id) = self.tcx.map.as_local_node_id(callsite.caller) {
+                    // Run copy propagation on the function to clean up any unnecessary
+                    // assignments from integration. This also increases the chance that
+                    // this function will be inlined as well
+                    debug!("Running copy propagation");
+                    let src = MirSource::from_node(self.tcx, id);
+                    MirPass::run_pass(&mut CopyPropagation, self.tcx, src, caller_mir);
+                };
+
+                true
+            }
+            kind => {
+                caller_mir[callsite.bb].terminator = Some(Terminator {
+                    source_info: terminator.source_info,
+                    kind: kind
+                });
+                false
+            }
+        }
+    }
+
+    fn cast_box_free_arg(&self, arg: Lvalue<'tcx>, ptr_ty: Ty<'tcx>,
+                         callsite: &CallSite<'tcx>, caller_mir: &mut Mir<'tcx>) -> Operand<'tcx> {
+        let arg = Rvalue::Ref(
+            self.tcx.mk_region(ty::ReErased),
+            BorrowKind::Mut,
+            arg.deref());
+
+        let ty = arg.ty(caller_mir, self.tcx).expect("Rvalue has no type!");
+        let ref_tmp = TempDecl { ty: ty };
+        let ref_tmp = caller_mir.temp_decls.push(ref_tmp);
+        let ref_tmp = Lvalue::Temp(ref_tmp);
+
+        let ref_stmt = Statement {
+            source_info: callsite.location,
+            kind: StatementKind::Assign(ref_tmp.clone(), arg)
+        };
+
+        caller_mir[callsite.bb]
+            .statements.push(ref_stmt);
+
+        let pointee_ty = match ptr_ty.sty {
+            ty::TyBox(ty) => ty,
+            ty::TyRawPtr(tm) | ty::TyRef(_, tm) => tm.ty,
+            _ => bug!("Invalid type `{:?}` for call to box_free", ptr_ty)
+        };
+        let ptr_ty = self.tcx.mk_mut_ptr(pointee_ty);
+
+        let raw_ptr = Rvalue::Cast(CastKind::Misc, Operand::Consume(ref_tmp), ptr_ty);
+
+        let cast_tmp = TempDecl { ty: ptr_ty };
+        let cast_tmp = caller_mir.temp_decls.push(cast_tmp);
+        let cast_tmp = Lvalue::Temp(cast_tmp);
+
+        let cast_stmt = Statement {
+            source_info: callsite.location,
+            kind: StatementKind::Assign(cast_tmp.clone(), raw_ptr)
+        };
+
+        caller_mir[callsite.bb]
+            .statements.push(cast_stmt);
+
+        Operand::Consume(cast_tmp)
+    }
+
+    fn make_call_args(&self, args: Vec<Operand<'tcx>>,
+                      callsite: &CallSite<'tcx>, caller_mir: &mut Mir<'tcx>) -> Vec<Operand<'tcx>> {
+        let tcx = self.tcx;
+        // FIXME: Analysis of the usage of the arguments to avoid
+        // unnecessary temporaries.
```

Comment:
That's the job of copy elimination. What you *could* do is not generate temporaries for every argument that does not contain a function-accessible pointer/index.